// @bubblewrap/core/dist/lib/features/EmptyFeature.js
"use strict";
/*
 * Copyright 2020 Google Inc. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmptyFeature = void 0;
class EmptyFeature {
    constructor(name) {
        this.buildGradle = {
            repositories: new Array(),
            dependencies: new Array(),
            configs: new Array(),
        };
        this.androidManifest = {
            permissions: new Array(),
            components: new Array(),
            applicationMetadata: new Array(),
            launcherActivityEntries: new Array(),
        };
        this.applicationClass = {
            imports: new Array(),
            variables: new Array(),
        };
        this.launcherActivity = {
            imports: new Array(),
            variables: new Array(),
            methods: new Array(),
        };
        this.delegationService = {
            imports: new Array(),
        };
        this.name = name;
    }
}
exports.EmptyFeature = EmptyFeature;
